### PR TITLE
[Multiple Hardwares][Rest-server] Support multiple hardwares in default scheduler 

### PR DESCRIPTION
**Pull request by hzy46**

### CODE REVIEW

File: src/rest-server/config/rest_server.py [hunk 1]
```diff
@@ -58,6 +74,22 @@ def run(self):
...
+
+        # default-computing-device-type only works in default scheduler.
+        # If hived scheduler is enabled, it will be ignored.
+        # It determines:
+        # (1) the resource name to be added in containers[*].resources.limits when we submit job.
+        # (2) how we count the gpu number.
+        # Users should place their computing device in `layout.yaml`.
+        # Currently we only support one computing device in default scheduler.
+        # Here we find all available computing devices from `layout.yaml`,
+        # and use the first one as the default computing device type.
```
Comment (abuccts) [inline src/rest-server/config/rest_server.py:86]: when changing `computing_device_types` from `set` to `list`, the order is undefined. It's probably not the one first added.
https://github.com/microsoft/pai/blob/8dbb5c3d04fa37c5a5529c0c208ea5bcb33e3024/src/rest-server/config/rest_server.py#L43
https://github.com/microsoft/pai/blob/8dbb5c3d04fa37c5a5529c0c208ea5bcb33e3024/src/rest-server/config/rest_server.py#L53-L54
https://github.com/microsoft/pai/blob/8dbb5c3d04fa37c5a5529c0c208ea5bcb33e3024/src/rest-server/config/rest_server.py#L89

File: src/rest-server/config/rest_server.py [hunk 2]
```diff
@@ -75,4 +107,9 @@ def validation_post(self, cluster_object_model):
         if 'kubernetes' not in cluster_object_model['layout'] or 'api-servers-url' not in cluster_object_model['layout']['kubernetes']:
             return False, 'kubernetes.api-servers-url is required'
 
+        if len(cluster_object_model['hivedscheduler']['config']) == 0:
```
Comment (abuccts) [inline src/rest-server/config/rest_server.py:110]: maybe `< 2`
https://github.com/microsoft/pai/blob/8dbb5c3d04fa37c5a5529c0c208ea5bcb33e3024/src/rest-server/deploy/rest-server.yaml.template#L66-L70

File: src/rest-server/src/models/v2/job/k8s.js [hunk 3]
```diff
@@ -589,6 +589,7 @@ const generateTaskRole = (
       },
     },
   };
+  console.log('frameworkTaskRole', JSON.stringify(frameworkTaskRole));
```
Comment (abuccts) [inline src/rest-server/src/models/v2/job/k8s.js:592]: remove

File: src/rest-server/config/rest_server.py (hunk 1 above)
Comment (hzy46) [reply src/rest-server/config/rest_server.py:89]: Done. Use another `list` together with `set`.

File: src/rest-server/src/models/v2/job/k8s.js (hunk 3 above)
Comment (hzy46) [reply src/rest-server/src/models/v2/job/k8s.js]: Done

File: src/rest-server/config/rest_server.py (hunk 2 above)
Comment (hzy46) [reply src/rest-server/config/rest_server.py]: Done

File: src/rest-server/config/rest_server.py [hunk 4]
```diff
@@ -38,6 +38,25 @@ def validation_pre(self):
...
+          for sku_name, sku_attrs in self.cluster_configuration['machine-sku'].items()
+          if 'computing-device' in sku_attrs
+      }
+      workers = list(filter(lambda elem: 'pai-worker' in elem and elem["pai-worker"] == 'true', self.cluster_configuration['machine-list']))
+      for worker in workers:
+        if worker['machine-type'] in machine_sku_to_device_type:
+          computing_device_type = machine_sku_to_device_type[worker['machine-type']]
+          if computing_device_type not in computing_device_type_set:
+            computing_device_type_set.add(computing_device_type)
+            computing_device_type_list.append(computing_device_type)
```
Comment (abuccts) [inline src/rest-server/config/rest_server.py:56]:
```suggestion
          if computing_device_type not in computing_device_type_list:
            computing_device_type_list.append(computing_device_type)
```

File: src/rest-server/config/rest_server.py [hunk 5]
```diff
@@ -38,6 +38,25 @@ def validation_pre(self):
 
         return True, None
 
+    # This function is used to find the all valid computing device types from `layout.yaml`
+    def find_computing_device_types(self):
+      computing_device_type_set = set()
```
Comment (abuccts) [inline src/rest-server/config/rest_server.py:43]: remove

Comment (hzy46) [reply src/rest-server/config/rest_server.py]: Done

File: src/rest-server/config/rest_server.py (hunk 4 above)
Comment (hzy46) [reply src/rest-server/config/rest_server.py]: Done

**Review by abuccts — APPROVED** [review]

**Pull request merged**